### PR TITLE
feat(payment): INT-7044 BlueSnapDirect: Add payment strategy for credit card

### PR DESCRIPTION
**Pull request by mauricio-sg**
## What? [INT-7044](https://bigcommercecloud.atlassian.net/browse/INT-7044)
Add a new mono repo 📦 for BlueSnap direct:

- This time we use [BlueSnap's hosted payment fields](https://developers.bluesnap.com/v8976-Tools/reference/hosted-payment-fields) and an additional field for cardholder name provided by BigCommerce.

## Why?
~All the logic will go here and will be added later. I want to avoid a big PR for an easier review. 😬~

To allow shoppers to pay with a credit card through BlueSnap.

## Testing / Proof
Hosted fields:

https://user-images.githubusercontent.com/4843328/218590902-a9b1af4e-a81f-47a6-928e-33b192e33ac4.mov

---

Purchase:

https://user-images.githubusercontent.com/4843328/225397822-d233235f-77d6-4f2f-b17a-951eac0dd205.mov

## Depends on
👉 #1810

@bigcommerce/checkout @bigcommerce/payments

[INT-7044]: https://bigcommercecloud.atlassian.net/browse/INT-7044?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by AlanMunozN — APPROVED** [review]

**Review by josesonora — APPROVED** [review]

File: packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import {
+    PaymentIntegrationService,
+    PaymentStrategy,
+} from '@bigcommerce/checkout-sdk/payment-integration-api';
+
+export default class BlueSnapDirectCreditCardPaymentStrategy implements PaymentStrategy {
+    constructor(private _paymentIntegrationService: PaymentIntegrationService) {}
+
+    execute(): Promise<void> {
```
Comment (animesh1987) [inline packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:9]: Hey @josesonora can you please add the rest of strategy as well. Can do it in a separate commit. Since most of the code is here already and feel will be better to move it in one chunk.

Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts]: Hey @animesh1987, the code is mine 😅. I added the rest of the code.

File: packages/credit-card-integration/src/credit-card-payment-strategy.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@ export default class CreditCardPaymentStrategy implements PaymentStrategy {
     protected _hostedForm?: HostedForm;
     protected _shouldRenderHostedForm?: boolean;
 
-    constructor(private _paymentIntegrationService: PaymentIntegrationService) {}
+    constructor(protected _paymentIntegrationService: PaymentIntegrationService) {}
```
Comment (animesh1987) [inline packages/credit-card-integration/src/credit-card-payment-strategy.ts:23]: What's the purpose of changing this?

File: packages/core/src/payment/payment-request-transformer.ts [hunk 3]
```diff
@@ -17,9 +17,12 @@ import { mapToInternalShippingOption } from '../shipping';
 
 import isVaultedInstrument, { isFormattedVaultedInstrument } from './is-vaulted-instrument';
 import Payment, {
+    CreditCardInstrument,
```
Comment (animesh1987) [inline packages/core/src/payment/payment-request-transformer.ts:20]: Same here, do we need to update this bit of code as part of creating this strategy?

File: packages/credit-card-integration/src/credit-card-payment-strategy.ts (hunk 2 above)
Comment (mauricio-sg) [reply packages/credit-card-integration/src/credit-card-payment-strategy.ts:23]: Just to make it visible to my subclass (BlueSnap). But, I can make my own variable and undo this change if you prefer.

File: packages/core/src/payment/payment-request-transformer.ts (hunk 3 above)
Comment (mauricio-sg) [reply packages/core/src/payment/payment-request-transformer.ts:20]: I'm afraid I don't understand this question. Do you mean the import thing or the whole file?

**Comment by mauricio-sg** [comment]
@animesh1987 let me know if you have more questions/requests ♻️

File: packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        const form = { ...creditCard.form };
+
+        if (!isHostedCardFieldOptionsMap(form.fields)) {
+            throw new InvalidArgumentError();
+        }
+
+        Reflect.set(form, 'fields', {
+            [HostedFieldType.CardName]: form.fields[HostedFieldType.CardName],
+        });
```
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:51]: To allow `creditCard.form.fields` to only contain a `CardName` field. Otherwise, the linter complains that fields are missing.

File: packages/core/src/payment/payment-request-transformer.ts [hunk 5]
```diff
@@ -166,24 +169,52 @@ export default class PaymentRequestTransformer {
...
+        }
+
+        return result;
+    }
+
+    private _transformToBlueSnapDirectInstrument({
+        nonce: pfToken,
+        hostedFormNonce,
+        ccName: cardHolderName,
+    }: NonceInstrument & CreditCardInstrument & WithHostedFormNonce): PaymentInstrument {
```
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/core/src/payment/payment-request-transformer.ts]: Done

File: packages/core/src/payment/payment-request-transformer.ts [hunk 6]
```diff
@@ -166,24 +169,52 @@ export default class PaymentRequestTransformer {
...
+                      ccCvv: values.cardCode,
+                      ccExpiry: this._cardExpiryFormatter.toObject(values.cardExpiry || ''),
+                      ccName: values.cardName || '',
+                      ccNumber: this._cardNumberFormatter.unformat(values.cardNumber || ''),
+                      hostedFormNonce: nonce,
+                  };
+
+        if (paymentMethod?.gateway === 'bluesnapdirect' && paymentMethod?.id === 'cc') {
+            return this._transformToBlueSnapDirectInstrument(
+                result as NonceInstrument & CreditCardInstrument & WithHostedFormNonce,
```
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/core/src/payment/payment-request-transformer.ts]: Done

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import { HostedFieldType } from '@bigcommerce/checkout-sdk/payment-integration-api';
+
+import { BlueSnapDirectHostedFieldTagId as HostedFieldTagId } from './types';
+
+export const HOSTED_FIELD_TYPES: {
```
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:5]: Done

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    [HostedFieldTagId.CardNumber]: HostedFieldType.CardNumber;
+    [HostedFieldTagId.CardExpiry]: HostedFieldType.CardExpiry;
+    [HostedFieldTagId.CardCode]: HostedFieldType.CardCode;
+} = Object.freeze({
```
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:9]: They are already `const`, so it's not possible to use `readonly`. Should I remove freeze? I added them to avoid mutations, `const` / `readonly` doesn't prevent the above.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:51]: Done. I declared a new type specific to BlueSnap.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (bc-peng) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:9]: As we are using typescript for the whole project, I think compile time checking is enough. Also, we got a unique type name here.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+        it('should load the Hosted Payment Fields SDK for sandbox', async () => {
+            await blueSnapDirectScriptLoader.load(true);
+
+            expect(scriptLoader.loadScript).toHaveBeenCalledWith(BlueSnapDirectSdkEnv.SANDBOX);
+        });
+
+        it('should fail to load the Hosted Payment Fields SDK', async () => {
+            jest.spyOn(scriptLoader, 'loadScript').mockImplementationOnce(() => {
+                Reflect.deleteProperty(window, 'bluesnap');
```
Comment (animesh1987) [inline packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts:46]: Do we really need to use Reflect here, and not use object methods?

Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts:46]: I need to delete de prop to simulate that the script was not loaded. `delete` operator is not allowed, the linter complains.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts]: Done, I removed `freeze`

**Comment by mauricio-sg** [comment]
@animesh1987 @bc-peng comments have been addressed, please let me know any questions in order to expedite the review of this PR, thanks. ♻️ cc: @david-intersys

File: packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts (hunk 9 above)
Comment (bc-peng) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts:46]: You may export a window interface from `is-bluesnap-direct-window.ts` as:
```
export interface BlueSnapDirectHostWindow extends Window {
    bluesnap?: BlueSnapDirectSdk;
}
```
In this file, at beginning:
```
jest.spyOn(scriptLoader, 'loadScript').mockImplementation(() => {
    (window as BlueSnapDirectHostWindow).bluesnap = blueSnapDirectSdkMock;

    return Promise.resolve();
});
```
And here:
```
jest.spyOn(scriptLoader, 'loadScript').mockImplementationOnce(() => {
    delete (window as BlueSnapDirectHostWindow).bluesnap;

    return Promise.resolve();
});
```

File: packages/core/src/hosted-form/hosted-form-options.ts [hunk 10]
```diff
@@ -38,6 +41,11 @@ export interface HostedStoredCardFieldOptionsMap {
     [HostedFieldType.CardNumberVerification]?: HostedStoredCardFieldOptions;
 }
 
+type BlueSnapDirectHostedCardFieldOptionsMap = Pick<
+    HostedCardFieldOptionsMap,
+    HostedFieldType.CardName
+>;
+
```
Comment (bc-peng) [inline packages/core/src/hosted-form/hosted-form-options.ts:48]: This file will be deprecated in favor of `packages/payment-integration-api/src/hosted-form/hosted-form-options.ts`. As a result, any modification is discouraged.

In addition, we wish to keep this file free of payment method specifics. Could you possibly put the BlueSnap material into its own package?

File: packages/core/src/hosted-form/hosted-form-order-data.ts [hunk 11]
```diff
@@ -17,7 +18,9 @@ export default interface HostedFormOrderData {
     config?: Config;
     order?: Order;
     orderMeta?: OrderMeta;
-    payment?: (HostedCreditCardInstrument | HostedVaultedInstrument) & PaymentInstrumentMeta;
+    payment?:
+        | ((HostedCreditCardInstrument | HostedVaultedInstrument) & PaymentInstrumentMeta)
+        | NonceInstrument;
```
Comment (bc-peng) [inline packages/core/src/hosted-form/hosted-form-order-data.ts:23]: This one appears to be payment method specific as well; could you kindly consider the Square approach?
https://github.com/bigcommerce/checkout-sdk-js/blob/b445659272efda3f381f1ed7332d8adef5da566d/packages/core/src/payment/strategies/square/square-form.ts#L154

Comment (bc-peng) [reply packages/core/src/hosted-form/hosted-form-order-data.ts:23]: You can import `NonceInstrument` from `@bigcommerce/checkout-sdk/payment-integration-api`, but you need to add it to the export first [here](https://github.com/bigcommerce/checkout-sdk-js/blob/b445659272efda3f381f1ed7332d8adef5da566d/packages/payment-integration-api/src/index.ts#L61).

File: packages/core/src/payment/payment-request-transformer.ts [hunk 12]
```diff
@@ -166,24 +168,54 @@ export default class PaymentRequestTransformer {
...
+            formattedPayload: {
+                credit_card_token: {
+                    token: JSON.stringify({
+                        pfToken,
+                        cardHolderName,
+                    }),
+                },
+                hostedFormNonce,
+            },
+        };
```
Comment (bc-peng) [inline packages/core/src/payment/payment-request-transformer.ts:218]: Similar to prior comments. How about putting this inside the Bluesnap package and using the payment integration API's `submitPayment(payment: Payment): Promise<PaymentIntegrationSelectors>` to submit Bluesnap required data?

File: packages/credit-card-integration/src/credit-card-payment-strategy.ts (hunk 2 above)
Comment (bc-peng) [inline packages/credit-card-integration/src/credit-card-payment-strategy.ts:23]:
```suggestion
    constructor(protected paymentIntegrationService: PaymentIntegrationService) {}
```
Makes sense to me, so we'd better remove `_` as it is no longer private.

Comment (bc-peng) [reply packages/credit-card-integration/src/credit-card-payment-strategy.ts:23]:
```suggestion
    constructor(protected paymentIntegrationService: PaymentIntegrationService) {}
```
Makes sense to me, so we'd better remove `_` as it is no longer private.

File: packages/payment-integration-api/src/hosted-form/hosted-form-options.ts [hunk 13]
```diff
@@ -40,6 +43,11 @@ interface HostedStoredCardFieldOptionsMap {
     [HostedFieldType.CardNumberVerification]?: HostedStoredCardFieldOptions;
 }
 
+type BlueSnapDirectHostedCardFieldOptionsMap = Pick<
+    HostedCardFieldOptionsMap,
+    HostedFieldType.CardName
+>;
```
Comment (bc-peng) [inline packages/payment-integration-api/src/hosted-form/hosted-form-options.ts:49]: Same here. We want to keep `payment-integration-api` independent of payment method-specific knowledge.

File: packages/credit-card-integration/src/credit-card-payment-strategy.ts (hunk 2 above)
Comment (mauricio-sg) [reply packages/credit-card-integration/src/credit-card-payment-strategy.ts:23]: Done.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts (hunk 9 above)
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts:46]: Done.

File: packages/core/src/hosted-form/hosted-form-options.ts (hunk 10 above)
Comment (mauricio-sg) [reply packages/core/src/hosted-form/hosted-form-options.ts:48]: I know, but since they are not fully decoupled yet, I had to modify both files. Otherwise, we get an error:

<pre>
ERROR in packages/core/src/payment/strategies/credit-card/credit-card-payment-strategy.ts(63,59)
        TS2345: Argument of type 'import("packages/<strong>payment-integration-api</strong>/src/hosted-form/<strong>hosted-form-options</strong>").default' is not assignable to parameter of type 'import("packages/<strong>core</strong>/src/hosted-form/<strong>hosted-form-options</strong>").default'.
<pre/>

File: packages/core/src/payment/payment-request-transformer.ts (hunk 12 above)
Comment (mauricio-sg) [reply packages/core/src/payment/payment-request-transformer.ts:218]: AFAIK it's not possible to retrieve the `cardHolderName` on the BlueSnap package because this value lives on another iframe. If I could access this value, the payment integration API's `submitPayment` method would be sufficient. Although I don't think it's a good idea.

File: packages/payment-integration-api/src/hosted-form/hosted-form-options.ts (hunk 13 above)
Comment (mauricio-sg) [reply packages/payment-integration-api/src/hosted-form/hosted-form-options.ts:49]: Please, see my previous comment.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts (hunk 9 above)
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-script-loader.spec.ts]: Undone.

File: packages/core/src/hosted-form/hosted-form-options.ts (hunk 10 above)
Comment (mauricio-sg) [reply packages/core/src/hosted-form/hosted-form-options.ts]: Undone.

File: packages/core/src/payment/payment-request-transformer.ts (hunk 12 above)
Comment (mauricio-sg) [reply packages/core/src/payment/payment-request-transformer.ts]: Undone.

**Comment by mauricio-sg** [comment]
@animesh1987 @bc-peng ♻️ I made the changes as discussed by DM.

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    [HostedFieldTagId.CardNumber]: HostedFieldType.CardNumber;
+    [HostedFieldTagId.CardExpiry]: HostedFieldType.CardExpiry;
+    [HostedFieldTagId.CardName]: HostedFieldType.CardName;
+    [HostedFieldTagId.CardCode]: HostedFieldType.CardCode;
+} = {
+    [HostedFieldTagId.CardNumber]: HostedFieldType.CardNumber,
+    [HostedFieldTagId.CardExpiry]: HostedFieldType.CardExpiry,
+    [HostedFieldTagId.CardName]: HostedFieldType.CardName,
+    [HostedFieldTagId.CardCode]: HostedFieldType.CardCode,
+};
```
Comment (bc-peng) [inline packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:15]:
```suggestion
export const BlueSnapHostedFieldType = {
    [HostedFieldTagId.CardNumber]: HostedFieldType.CardNumber,
    [HostedFieldTagId.CardExpiry]: HostedFieldType.CardExpiry,
    [HostedFieldTagId.CardName]: HostedFieldType.CardName,
    [HostedFieldTagId.CardCode]: HostedFieldType.CardCode,
};
```

Comment (bc-peng) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:15]: Is there any reason you put type information for a const?

File: packages/payment-integration-api/src/errors/invalid-hosted-form-value-error.ts
Comment (bc-peng) [inline packages/payment-integration-api/src/errors/invalid-hosted-form-value-error.ts:1]: Can we use this existing error type?
https://github.com/bigcommerce/checkout-sdk-js/blob/master/packages/payment-integration-api/src/errors/payment-invalid-form-error.ts

File: packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts (hunk 14 above)
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-constants.ts:15]: I use `BlueSnapHostedFieldType` to map types between BlueSnap and BigCommerce, so I need to be this specific. If I commit your suggestion TS would infer that any key of `BlueSnapHostedFieldType` returns a `HostedFieldType`:

```
const BlueSnapHostedFieldType: {
    ccn: HostedFieldType;
    exp: HostedFieldType;
    noc: HostedFieldType;
    cvv: HostedFieldType;
}
```

This gives me errors because `HostedFieldType.CardCodeVerification` and `HostedFieldType.CardNumberVerification` does not exist on my mappings. But, if I explicitly exclude them everything is fine:

```
const BlueSnapHostedFieldType: {
    ccn: HostedFieldType.CardNumber;
    exp: HostedFieldType.CardExpiry;
    noc: HostedFieldType.CardName;
    cvv: HostedFieldType.CardCode;
}
```

Sorry, I haven't found a more elegant way to do this mapping, please let me know if you have a better idea.

File: packages/payment-integration-api/src/errors/invalid-hosted-form-value-error.ts
Comment (mauricio-sg) [reply packages/payment-integration-api/src/errors/invalid-hosted-form-value-error.ts:1]: Done.

**Review by bc-peng — APPROVED** [review]
Nice work, thank you for your patience.

**Review by animesh1987 — APPROVED** [review]

File: packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        const {
+            config: { testMode },
+            clientToken,
+        } = (
+            await this._paymentIntegrationService.loadPaymentMethod(gatewayId, {
+                params: { method: methodId },
+            })
+        ).getPaymentMethodOrThrow(methodId, gatewayId);
```
Comment (animesh1987) [inline packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:42]:
```suggestion
        const {
            config: { testMode },
            clientToken,
        } = 
            await this._paymentIntegrationService.loadPaymentMethod(gatewayId, {
                params: { method: methodId },
            })
        .getPaymentMethodOrThrow(methodId, gatewayId);
```

Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:42]: I need to await the promise first otherwise, I get:
`Property 'getPaymentMethodOrThrow' does not exist on type 'Promise<PaymentIntegrationSelectors>'.`

Comment (animesh1987) [reply packages/bluesnap-direct-integration/src/bluesnap-direct-credit-card-payment-strategy.ts:42]: Then in that case maybe separate them in two lines?

File: packages/bluesnap-direct-integration/src/bluesnap-direct-name-on-card-input.ts [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            CSSStyleDeclaration,
+            'color' | 'fontFamily' | 'fontSize' | 'fontWeight'
+        > = {
+            color: styles.color,
+            fontFamily: styles['font-family'],
+            fontSize: styles['font-size'],
+            fontWeight: styles['font-weight'],
+        };
+
+        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
Comment (animesh1987) [inline packages/bluesnap-direct-integration/src/bluesnap-direct-name-on-card-input.ts:111]: Can we do this using type guards?

**Comment by JelopezBC** [comment]
💚 Tested!
Story has been validated and is working as expected. 

Please review the evidence attached.
https://drive.google.com/drive/folders/1hfviefrzCWWf22EW4mH6edwVA-IhmZ7Y?usp=share_link
